---
title: "Simulation of Datasets with Splatter"

subtitle: "ScMaSigPro Supplementary Material-II"

author: "Priyansh Srivastava, ... Ana Conesa"

date: "`r Sys.Date()`"

knit: (function(inputFile, encoding) {
     rmarkdown::render(inputFile, encoding=encoding,
     output_file=file.path("../Supplementary_Material/02_Simulation_with_Splatter.pdf"))})
     
output:
 pdf_document:
     latex_engine: lualatex
     citation_package: natbib

bibliography: Refs/references.bib

header-includes:
  - "\\usepackage{fontspec}"
  - "\\setmainfont{TeX Gyre Termes}"
  - "\\setlength{\\parindent}{1em}"    # Indentation width, change 1em to your preference
  - "\\setlength{\\parskip}{1em}"      # Space between paragraphs

---

# Introduction
To evaluate the effectiveness of the scMaSigPro in controlling the False Positive Rates (FPR) and False Negatives Rates (FNR), we have evaluated its application on a wide variety of simulated scRNA-datasets. As in simulated scRNA-datasets, the ground truth is well established; therefore, simulations serve as the basis to benchmark the accuracy and precision of the method.

We have used Splatter, which uses a Gamma-Poisson distribution to simulate scRNA-Seq data [@zappia_splatter_2017]. We simulated Splatter's ```splatSimulate(method = "paths")``` function to simulate a differentiation process where one cell type changes into another [@luke_zappia_introduction_2020]. Splatter approximates this process by simulating a series of steps between two groups and randomly assigning each cell to a **Step**. Since Pseudotime values are arbitrary, **"Step"** can be treated as Pseudotime. Additionally, just like Pseudotime, the **steps** simulated by Splatter also starts from 0, denoting immature cells first with a pseudotime of 0, followed by the cells in the transitional stage pseudotime > 0, and finally, the mature cells.  [@deconinck_recent_2021].

## General Parameters for Datasets
We simulated bifurcating topologies (One cell type divides into two different cell types) with different parameters. $5000$ features/Genes were simulated across $3000$ cells/samples for all the simulations along two **paths/groups**. This produces a scRNA-Dataset where $~1500$ cells exist in each **paths/groups** following **Steps/Pseudotime** $0$ to $1500$. This represents an ideal case in which one cell state is associated with one value of **Step/Pseudotime**. This reflects an ideal sequencing experiment in which all the cell states are captured and sequenced, and the inferred **Steps/Pseudotime** orders all the cells in the native biological order. See Figure \ref{fig:ideal_bifurcating_path}.

\begin{figure}[!htbp]
\centering
\includegraphics[width=\textwidth]{Images/Fig1_Simulation_with_Splatter.png}
\caption{Ideal representation of native biological order. (A) Principal Components of bifurcating trajectory, where each cell is coloured by the associated Steps/Pseudotime. The length of the one path/group $~1500$ where each cell/sample is associated with one Steps/Pseudotime. (B) Each is coloured to represent a bifurcating trajectory}
\label{fig:ideal_bifurcating_path}
\end{figure}

## Annotation of Differential Genes (Ground Truth)

Splatter, simulates the differential expression by simulating fold-change per **path/group** which reflect the resulting change at the end of the **path/Group** in relation to the start i.e. $Gene_{i_{base}}$ . The final expression of $gene_i$ in $Path_{i}$ is the product of $Path_{i_{foldChange}} * Gene_{i_{base}}$. The effective change in expression across the **Steps/Pseudotime** can be obtained by taking the difference between the start $Gene_{i_{base}}$ and end $Path_{i_{foldChange}} * Gene_{i_{base}}$. If the difference is $+ve$ then the gene is over-expressed in the **path/group** along the **Step/Pseudotime** and under-expressed if the difference is $-ve$.

Therefore, we annotated the genes as follows,

1. If the effective change in expression is more than 0: **Differentially Expressed Along the Lineage**
    - If the difference is more than 2 units: **High-Fold change + Differentially Expressed**
    - If the difference is less than 2 units: **Low-Fold change + Differentially Expressed**
2. If the effective change in expression is equal to 0: **Non-Differentially Expressed Along the Lineage**

Given a bifurcating trajectory a gene which is differentially expressed can following 4 patterns along the trajectory, as discussed in figure  \ref{fig:what_is_differentially_expressed}. We further annotated the simulated genes to identify which of the 4 possible patterns each gene follows.


\begin{figure}[!htbp]
\centering
\includegraphics[width=\textwidth]{Images/Fig2_What_is_differentially_Expressed.png}
\caption{Four possible patterns of the gene along a bifurcating trajectory. (A) A Gene showing a similar change in expression towards the end of the trajectory of both the lineages/paths can be both over/under-expressed, but the effective change will be similar in both the lineage/paths. (B) A Gene showing the opposite change in expression towards the end of the trajectory of both the lineages/paths. (C) A Gene showing the change in expression towards the end of the trajectory of only one lineage/path, can be both over/under-expressed but change in only path relative to the other. (D) A gene which is not differentially expressed.}
\label{fig:what_is_differentially_expressed}
\end{figure}


## Specific Parameters for benchmarking

### 1. Zero-Inflation/Drop-out


# References

